### PR TITLE
Migrate from TSLint to ESLint

### PR DESCRIPTION
**Pull request by zweimach**
I made this changes last year but haven't had the courage to open a PR.
Following issue #2010 made by @calvinf, I encourage myself to open this.
Because TSLint was deprecated last year, I made this commit to migrates from TSLint to ESLint with support for TypeScript.
I already fixed some of the lint rules violation created by running ESLint against existing files.
Please kindly review my PR. Thank you.

### CODE REVIEW

File: src/esprima.ts [hunk 1]
```diff
@@ -94,11 +94,10 @@ export function parseScript(code: string, options, delegate) {
 export function tokenize(code: string, options, delegate) {
     const tokenizer = new Tokenizer(code, options);
 
-    let tokens;
-    tokens = [];
+    const tokens: any = [];
 
     try {
-        while (true) {
+        for (; ;) {
```
Comment (ariya) [inline src/esprima.ts]: Can we put a rule option that permits `while (true)`?
IMO `for(;;)` is far less readable (it does not telegraph the semantic of "forever").

**Comment by ariya** [comment]
Thanks @zweimach! Can you rebase it on the latest master.
(Note: at some point,  better to squash this PR before it is merged)

**Comment by zweimach** [comment]
@ariya feel free to squash those commits, and thanks for the review.
Glad to be of help.

File: src/scanner.ts [hunk 2]
```diff
@@ -1109,7 +1110,7 @@ export class Scanner {
                 .replace(/\\u\{([0-9a-fA-F]+)\}|\\u([a-fA-F0-9]{4})/g, ($0, $1, $2) => {
                     const codePoint = parseInt($1 || $2, 16);
                     if (codePoint > 0x10FFFF) {
-                        self.throwUnexpectedToken(Messages.InvalidRegExp);
+                        this.throwUnexpectedToken(Messages.InvalidRegExp);
```
Comment (ariya) [inline src/scanner.ts]: Hmm, why is this changed?

File: src/jsx-parser.ts [hunk 3]
```diff
@@ -35,20 +35,23 @@ function getQualifiedElementName(elementName: JSXNode.JSXElementName): string {
     let qualifiedName;
 
     switch (elementName.type) {
-        case JSXSyntax.JSXIdentifier:
+        case JSXSyntax.JSXIdentifier: {
```
Comment (ariya) [inline src/jsx-parser.ts]: Is there also a rule/option somewhere that minimize this unnecessary change?

File: src/scanner.ts (hunk 2 above)
Comment (zweimach) [reply src/scanner.ts]: That is a recommended lint rule of [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin),
You can read about that on [@typescript-eslint/no-this-alias](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-type-alias.md),
and here we use an arrow function expression so aliasing `this` is pretty much useless.

File: src/jsx-parser.ts (hunk 3 above)
Comment (zweimach) [reply src/jsx-parser.ts]: No [there isn't](https://eslint.org/docs/rules/no-case-declarations). Do you want me to revert that one?

File: src/scanner.ts (hunk 2 above)
Comment (ariya) [reply src/scanner.ts]: That makes sense! Thank you for the clarification.

File: src/jsx-parser.ts (hunk 3 above)
Comment (ariya) [reply src/jsx-parser.ts]: Yes, please!

File: package.json [hunk 4]
```diff
@@ -75,11 +77,11 @@
   ],
   "scripts": {
     "check-version": "node test/check-version.js",
-    "tslint": "tslint src/*.ts",
+    "lint-code": "eslint src/*.ts",
```
Comment (calvinf) [inline package.json]: I'd call it `lint`, but that's personal preference.

Comment (zweimach) [reply package.json]: I choose `lint-code` because there is a script called `format-code`.
@ariya what do you think?

**Comment by ariya** [comment]
Merged. Thank you @zweimach!
